ls1intum/Artemis
# Feature/update structural tests

### Checklist

### Motivation and Context
Programming Exercises in Java: Improve the structural tests by allowing to define annotations in the test.json structural oracle.
Improve the oracle generator:
1. Add the detection of annotation changes
2. Take more properties of attributes, methods and constructors into account when comparing template and solution files to detect more changes.

### Description
I updated the structural tests and the test.json generator.

### Steps for Testing
1. Log in to Artemis
2. Navigate to Course Administration and into a Test Course
3. Create a new programming exercise with Java as programming language
4. Add some annotations into the solution repository for a class, attribute, method and/or constructor (make sure that the retention policy of the annotation is runtime: the easiest is to define your own custom annotation)
5. Commit and push the changes into the solution repository
6. Navigate into the programming exercise detail page, e.g. `course/10/programming-exercise/660/view`
7. Click on `Update Structure Test Oracle`. This takes a few seconds and should then automatically trigger a commit + push in the test repo (Look for a green or red alert at the top of the page.)
8. Check if the annotations are recognized correctly in the changes in the test repos `test.json` file (in the new commit)
9. Check if the template build (0%) and solution build (100%) are still as desired
10. Delete the programming exercise with all participations after finishing the test
